pkg/cloudprovider/provider/aws/provider.go

```diff
@@ -89,6 +91,11 @@ var (
 			owner: "099720109477",
 		},
 	}
+
+	// cacheLock protects concurrent cache misses against a single key. This usually happens when multiple machines get created simultaneously
+	// We lock so the first access updates/writes the data to the cache and afterwards everyone reads the cached data
+	cacheLock = &sync.Mutex{}
+	cache     = gocache.New(5*time.Minute, 5*time.Minute)
```

Comment:
Not sure about how I feel using a library for this. IMHO we should:

* Move the whole ami lookup into the `AddDefaults` to have it written into the `spec` to avoid the possibility that different machines with the same config from the same `MachineSet` end up using different AMIs 
* Add a validation step for the ami that checks if it exists into `Validate`
* Cache the validation for the whole `.Spec` by using a `map[string]bool` that holds a `hash` of the `.spec` and a `bool` that indicates success/failure

Reply:
No, it covers all of that, `Machine` will still get defaulted and validated if its not owned by a `MachineSet`. My point is, that I would like to keep the amount of caches we built as low as possible, hence only one that validates the whole of a `MachineSpec`